### ISSUE TITLE
Documenting the semantics of tables with pattern indices

### ISSUE THREAD

**Issue by vpax** [zeek/zeek#1561]
https://docs.zeek.org/en/current/script-reference/types.html#type-table mentions that tables can be indexed using patterns, such as `table[pattern] of T`, but it doesn't describe the associated semantics.  As best as I can tell, these aren't documented anywhere other than implicitly in test suite tests.

Also - though maybe this should remain a separate discussion - from some brief testing, the semantics seem a bit weird.  As best as I can tell, all you can do with such tables is iterate over them; you can't index them or use the `in` operator, even though the analogous `table[subnet] of T` type does support those operations.  Attempts to do either of those operations using either `string` or `pattern` values results in type errors.

**Comment by ckreibich** [comment]
Vern I'm seeing somewhat different things so let's sort it all out ... in the docs I actually see that tables do _not_ support patterns:

>  The index type cannot be any of the following types: pattern, table, set, vector, file, opaque, any.

I suspect this got overlooked when we expanded pattern support (in 3.1, iirc).

Now, indexing a table with a pattern [works here](https://try.zeek.org/#/tryzeek/saved/509264). Did your construct differ somehow? Was it an older Zeek version? (I didn't try pre-4.0).

I can confirm that `in` [does not work](https://try.zeek.org/#/tryzeek/saved/509265). Looking at the code, the `in` operator indeed requires a string as the right-hand type if the left is a pattern.

So at a minimum I'm going to:
- go over that list of supposedly unsupported types and see if any others actually work, and expand the docs accordingly
- add wording that explains the current, limited semantics

Did you encounter any other limitations around `table[pattern]`?

I haven't explored how hard it would be to expand `in` for `table[pattern]`, but I think we should give it a go.

**Comment by vpax** [comment]
Thanks, @ckreibich !  I misread that phrase during my search as saying that support _is_ provided.  I got down this path, however, due to `testing/btest/language/table-pattern-index.zeek`, whose presence clearly indicates an expectation that it would work.  I then tried the `in` operator, and indexing with a `string` (seems like a natural use-case, analogous to `addr`/`subnet`), but evidently didn't try directly indexing with a `pattern` (hard to picture how that would ever actually be of interest!).

I think the semantics are pretty weird, anyway, so would be inclined to remove the functionality rather than see how to make it work better.  The one use-case (present in the test suite entry) is really the need to _pair_ items, which I'd rather see done with a language feature oriented around tuples, rather than this quasi-abuse of `table` types.

**Comment by grigorescu** [comment]
Reading this, I'm recalling some conversations from years ago about this. I believe that not supporting string lookups in tables of patterns was a deliberate design choice. Patterns can introduce ambiguity regarding which element should be returned in some cases.

`table[subnet]` does the right thing, namely it returns the most specific match (i.e. the largest numerical CIDR range). We could try to adopt this to patterns as well (e.g. "Return the pattern with the longest match."), but if I have a table with elements `[/.bc/, /a.c/, /ab./]`, `t["abc"]` is undefined.

I would actually argue that while accessing a pattern table via a string would result in undefined behavior, the `in` operator would make sense -- do any of the patterns in this table match the string? If the user wants to know which one(s) do, they'd have to iterate through the table, which would be inconvenient.

**Comment by vpax** [comment]
Point taken re the semantics being more clear for `table[subnet] of T`.

Regarding avoiding iterating through the table, I don't think under the hood we could (easily) implement anything other than such iteration anyway.  Also, what's the utility of an `in` operator if you can't then access the match?  Might as well use `x in p` for `p` being the disjunction of all of the patterns of interest, rather than putting them in a table, if I'm following your example correctly.

**Comment by grigorescu** [comment]
Yeah, I've been trying to come up with a use case that's begging for this, and haven't been able to.

policy/protocols/http/detect-webapps.zeek ties a regex to the name of a webapp, but it uses the signature framework instead of patterns.

http/software.zeek is a bit more interesting: https://github.com/zeek/zeek/blob/52b3e159a415fb3acd584f95aa64e2d2a9c0b282/scripts/policy/protocols/http/software.zeek#L17-L18

I could easily see someone could extend that such that a given user agent is ignored only for specific subnets, or ports, or HTTP methods. However, I think the correct approach would be a `set[pattern, subnet]` rather than a table.

I do think that the `in` operator would be useful for sets of patterns. As a side-effect, it would also work for tables -- because it CAN, but its usage might be an anti-pattern.

In terms of implementation, my (naive) thought was that whenever the table gets updated, a pattern gets built which is just all the individual patterns `OR`-ed together. An `in` just matches against that pattern. I'm assuming that these tables would get updated very rarely, and the performance hit would be acceptable, but I suppose someone could go crazy with expiring patterns.

**Comment by ckreibich** [comment]
Seems there are two parts here: one is whether tables should support indexing via patterns, and the other is whether/how this ability should in turn extend to indexing such a table with strings, in a parallel to the subnet logic. I wouldn't naturally expect the latter to work. Some of this might also encroach on paraglob terrain. But I don't see why indexing directly with a pattern should not work, since we've long been able to generate patterns dynamically in the language.

There's one use case this reminded me of ... signature learning, where during the learning process you associate state with lots of in-flux patterns. Hardly common, for sure. Also, Vern, this is probably an instance of what you mean by pairing patterns with state. But you'd still want to find that state somehow, and it seems patterns could be the intuitive way to do it.

A higher-level point for me is that each unsupported or partially supported scenario makes for a potential surprise while working with the language. We often have hard technical reasons for not supporting something ... is this one of them? If this is a matter of either pulling indexing with patterns, or adding support for `in` so it's complete, I'd vote for the latter.

**Comment by ckreibich** [comment]
Btw, for the other types cautioned in the docs indexing does indeed not work. For tables, sets, and vector as index types you get away with defining the table type, but when trying to index, you get a 

    internal error: bad index type in CompositeHash::ComputeSingletonHash
    Aborted (core dumped)

For file, any, and opaques you get a more polite `index type containing '...' is not supported` error message.

**Comment by rsmmr** [comment]
> Some of this might also encroach on paraglob terrain.

I was just about to mention paraglob: the motivation for that was of course to support indexing with a string into a large set of patterns efficiently (and dynamically, with the set changing during runtime). There was always the question if we could/should add the syntactic sugar to hide paraglob behind table operators but the hasn't seemed worth it so far (and it would be globs, not full regexps anyways). But either way, I'd say that use case is pretty well covered through paraglob. 

Re/ the question whether to support indexing with patterns themselves (rather than strings), that would probably be easy to do, even though not sure how useful it really is; but if we want it just for consistency, fine. :)
